### PR TITLE
Sort new fields in lexicographic order during update

### LINKED ISSUE #2038: Sort document fields in lexicographic/numeric order
### What should be done?

https://www.mongodb.com/docs/manual/reference/operator/update/#behavior

* `"bar"` comes before `"foo"`
* `"7"` comes before `"42"`
* `"bar.7"` comes before `"bar.42"`

### Where?

https://github.com/FerretDB/FerretDB/blob/102b27d65fc2a8f689fb416eebc86bcffc842116/internal/types/document.go#L363

### Definition of Done

- `SortFieldsByKey` added to places where it should be;
- all handlers updated;
- unit tests added/updated;
- integration/compatibility tests added/updated;
- spot refactorings done;
- user documentation updated.

### PR DESCRIPTION
**Pull request by wazir-ahmed**
# Description
Closes #2038.

## Readiness checklist
- [ ] I added/updated unit tests (and they pass).
- [x] I added/updated integration/compatibility tests (and they pass).
- [x] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [ ] I ensured that PR title is good enough for the changelog.
- [ ] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [ ] I marked all done items in this checklist.

### CODE REVIEW

File: internal/handler/common/update_array_operators.go [hunk 1]
```diff
@@ -120,366 +103,353 @@ func checkUnsuitableValueError(doc *types.Document, key string, path types.Path)
...
-		array, ok := val.(*types.Array)
-		if !ok {
-			return false, handlererrors.NewWriteErrorMsg(
-				handlererrors.ErrBadValue,
-				fmt.Sprintf(
-					"The field '%s' must be an array but is of type '%s' in document {_id: %s}",
-					key, handlerparams.AliasFromType(val), types.FormatAnyValue(must.NotFail(doc.Get("_id"))),
-				),
-			)
+	if !doc.HasByPath(path) {
```
Comment (chilagrow) [inline internal/handler/common/update_array_operators.go:362]: Should it create a new array and set it? The logic appears to have changed.

File: integration/findandmodify_compat_test.go [hunk 2]
```diff
@@ -422,34 +422,47 @@ func TestFindAndModifyCompatUpdateCurrentDate(t *testing.T) {
 				{"query", bson.D{{"_id", "datetime"}}},
 				{"update", bson.D{{"$currentDate", 1}}},
 			},
+			providers:  []shareddata.Provider{shareddata.DateTimes},
```
Comment (chilagrow) [inline integration/findandmodify_compat_test.go:425]: Could you enlighten why we only test for datetime provider? We should be testing for all providers.

**Review by chilagrow — COMMENTED** [review]
Thanks for your contribution :hugs:

This is a huge change, we are taking time to review it carefully. Meanwhile I have a few questions :smiley:

File: integration/findandmodify_compat_test.go (hunk 2 above)
Comment (wazir-ahmed) [reply integration/findandmodify_compat_test.go:425]: We are having a filter of `{_id: "datetime"}` in these test cases. What is use of running it against all the providers? We will be reducing test execution time by limiting the providers whenever we use a specific `_id` filter in test cases.

File: internal/handler/common/update.go [hunk 3]
```diff
@@ -350,653 +351,570 @@ func processUpdateOperator(command string, doc, update *types.Document, upsert b
...
+	if doc.HasByPath(path) {
+		docValue := must.NotFail(doc.GetByPath(path))
+		if types.Identical(setValue, docValue) {
+			return false, nil
 		}
+	}
 
-		changed = true
+	// we should insert the value if it's a regular key
+	// if setOnInsert && path.Len() > 1 {
```
[the remark this answers is not in the document]
Comment (wazir-ahmed) [reply internal/handler/common/update.go:421]: > If input has dot notation, can we successfully set path?

Yes, we can. I have added a compat test cases for it - `TestUpdateFieldOrder/NestedFieldDotNotation`

> Why is this condition commented out?

I forgot to remove this block after testing. I will remove this in the next commit.

File: internal/handler/common/update.go [hunk 4]
```diff
@@ -350,653 +351,570 @@ func processUpdateOperator(command string, doc, update *types.Document, upsert b
...
-			case "date":
-				doc.Set(field, now)
-				changed = true
-			}
+	switch value := value.(type) {
+	case *types.Document:
+		// ignore error, error cases were validated in validateCurrentDateExpression
+		dateType, _ = value.Get("$type")
+	case bool:
+		dateType = "date"
```
[the remark this answers is not in the document]
Comment (wazir-ahmed) [reply internal/handler/common/update.go:818]: The `dateType` refers to BSON type - `Date` or `timestamp`. In MongoDB, if `bool` value is passed for a specific key with `$currentDate` operator, then current date will be set as `Date` type. That is what this line is referring to.

Perhaps, `dateType` name is confusing. I will rename it `setValType` and add comments explaining the behaviour of `$currentDate` operator. Is that fine? Or you would still prefer the previous implementation? Because I changed the prev impl, because it was confusing to me with nested switch cases. :sweat_smile:

File: internal/handler/common/update_array_operators.go (hunk 1 above)
Comment (wazir-ahmed) [reply internal/handler/common/update_array_operators.go:362]: Yes, the logic is changed. MongoDB doesn't add an array for non-existent fields in case of `$pop`, `$pull` and `$pullAll` operators.

We had this compatability bug before and it was not visible in our tests.
1. As per the previous logic (in all three operators), we will add an array for non-existent fields to the `doc`, but the `changed` flag will be `false` when function returns :upside_down_face:
2. In case an existing document, there won't be any compat issue, because we won't update the DB backends if `changed` is false.
3. But in case of upsert, where we ignore the `changed` flag and always push contents of `doc` pointer to DB backends, we will face the issue.
4. We didn't have a test case for [upsert + non-existent field] scenario, so this issue was not visible to us.

I have added compat test cases (`NonExistentFieldUpsert`) for all three operators and also modified `testUpdateCompat()` to compare the inserted documents, in case of upsert. :cowboy_hat_face:

File: internal/handler/common/update.go [hunk 5]
```diff
@@ -1361,10 +1276,8 @@ func validateCurrentDateExpression(command string, update *types.Document) error
 					)
 				}
 			}
-			currentDateType, err := setValue.Get("$type")
-			if err != nil { // ok, default is date
-				continue
-			}
+
+			currentDateType, _ := setValue.Get("$type")
```
Comment (chilagrow) [inline internal/handler/common/update.go:1280]: Why was this changed? What happens if `$type` isn't set?

File: integration/findandmodify_compat_test.go (hunk 2 above)
Comment (chilagrow) [reply integration/findandmodify_compat_test.go:425]: The reason why we run all providers is often one of providers catch some issue, particularly related to array filters. You are right, for this case datetime provider should be enough

File: internal/handler/common/update.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handler/common/update.go]: Ah got it, you wanted to avoid nested switch case. Looks fine now :hugs:

File: internal/handler/common/update.go [hunk 6]
```diff
@@ -350,653 +351,567 @@ func processUpdateOperator(command string, doc, update *types.Document, upsert b
...
+// If the document was changed it returns true.
+func processSetFieldExpression(command string, doc *types.Document, setKey string, setValue any, setOnInsert bool) (bool, error) {
+	if setOnInsert {
+		// $setOnInsert do not set null and empty array value.
+		if _, ok := setValue.(types.NullType); ok {
+			return false, nil
 		}
 
-		// we should insert the value if it's a regular key
-		if setOnInsert && path.Len() > 1 {
```
[the remark this answers is not in the document]
Comment (wazir-ahmed) [reply internal/handler/common/update.go:386]: This block is incorrect. In the case of `setOnInsert`, if we have a key with dot notation, we have to create the appropriate path and set the value. That is the behavior in MongoDB. I came to know about this, after `TestUpdateFieldOrder/NestedFieldDotNotation` case failed during my local testing.

File: internal/handler/common/update.go [hunk 7]
```diff
@@ -350,653 +351,567 @@ func processUpdateOperator(command string, doc, update *types.Document, upsert b
...
+		if err == iterator.ErrIteratorDone { //nolint:errorlint // only ErrIteratorDone could be returned
+			break
+		}
 
-		if setOnInsert {
-			// $setOnInsert do not set null and empty array value.
-			if _, ok := setValue.(types.NullType); ok {
-				continue
-			}
+		opDoc := opVal.(*types.Document)
```
[the remark this answers is not in the document]
Comment (wazir-ahmed) [reply internal/handler/common/update.go:368]: Yes. The validation happens in `ValidateUpdateOperators()`. For all update operator keys, the value should be a document.

File: internal/handler/common/update.go [hunk 8]
```diff
@@ -350,653 +351,567 @@ func processUpdateOperator(command string, doc, update *types.Document, upsert b
...
-	setDocKeys := setDoc.Keys()
-	sort.Strings(setDocKeys)
+	iter := update.Iterator()
+	defer iter.Close()
 
-	for _, setKey := range setDocKeys {
-		setValue := must.NotFail(setDoc.Get(setKey))
+	for {
+		operator, opVal, err := iter.Next()
+		if err == iterator.ErrIteratorDone { //nolint:errorlint // only ErrIteratorDone could be returned
```
[the remark this answers is not in the document]
Comment (wazir-ahmed) [reply internal/handler/common/update.go:364]: `iter` is a `documentIterator` (we are creating it from a doc, few lines above), and `documentIterator`s can only return `ErrIteratorDone`. Do you think, we still have to check for other error types?

File: internal/handler/common/update.go (hunk 5 above)
Comment (wazir-ahmed) [reply internal/handler/common/update.go:1280]: The previous implementation is incorrect. When `setValue` is a document, then `$type` field is mandatory. If not set, we should throw an error (as done in L1284). Added new test case - `TestUpdateFieldCompatCurrentDate/NoType` to verify this.

File: internal/handler/common/update.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handler/common/update.go:364]: Yes that is true, let's leave it as it is, the suggestion was for keeping the consistency across our code base. :hugs:

**Review by chilagrow — APPROVED** [review]
Huge thanks for your contribution :hugs:

**Review by AlekSi — APPROVED** [review]

**Comment by AlekSi** [comment]
Thank you!
